File: skellig-connection-http/src/main/java/org/skellig/connection/http/model/HttpResponse.java
package org.skellig.connection.http.model;

import java.util.HashMap;
import java.util.Map;

public final class HttpResponse {

    private int statusCode;
    private Map<String, String> headers;
    private String body;

    private HttpResponse(int statusCode, Map<String, String> headers, String body) {
        this.statusCode = statusCode;
        this.headers = headers;
        this.body = body;
    }

    public int getStatusCode() {
        return statusCode;
    }

    public Map<String, String> getHeaders() {
        return headers;
    }

    public String getBody() {
        return body;
    }

    public static class Builder {

        private int statusCode;
        private Map<String, String> headers = new HashMap<>();
        private String body;

        public Builder withBody(String body) {
            this.body = body;
            return this;
        }

        public Builder withHeaders(Map<String, String> theHeaders) {
            this.headers = theHeaders;
            return this;
        }

        public Builder withStatusCode(int code) {
            statusCode = code;
            return this;
        }

        public HttpResponse build() {
            return new HttpResponse(statusCode, headers, body);
        }

    }
}

File: skellig-test-step-processing-http/src/main/kotlin/org/skellig/teststep/processor/http/model/HttpResponse.kt
package org.skellig.teststep.processor.http.model

/**
 * HttpResponse represents an HTTP response received from a server.
 *
 * @property statusCode The HTTP status code of the response.
 * @property headers The headers of the response represented as a map.
 * @property body The body of the response represented as a string.
 */
class HttpResponse(val statusCode: Int,
                   val headers: Map<String, String>,
                   val body: String?) {

    override fun toString(): String {
        return "statusCode: $statusCode" +
                (if (headers.isNotEmpty()) ("\nheaders: $headers") else "") +
                (body?.let { ("\nbody: $body") }?:"")
    }

    class Builder {

        private var statusCode = 0
        private var headers = mapOf<String, String>()
        private var body: String? = null

        fun withBody(body: String?) = apply {
            this.body = body
        }

        fun withHeaders(theHeaders: Map<String, String>) = apply {
            headers = theHeaders
        }

        fun withStatusCode(code: Int) = apply {
            statusCode = code
        }

        fun build(): HttpResponse {
            return HttpResponse(statusCode, headers, body)
        }
    }
}